bundles/org.openhab.binding.mycroft/src/main/java/org/openhab/binding/mycroft/internal/MycroftHandler.java

```diff
@@ -0,0 +1,234 @@
+import org.openhab.binding.mycroft.internal.api.MessageType;
+import org.openhab.binding.mycroft.internal.api.MycroftConnection;
+import org.openhab.binding.mycroft.internal.api.MycroftConnectionListener;
+import org.openhab.binding.mycroft.internal.api.MycroftMessageListener;
+import org.openhab.binding.mycroft.internal.api.dto.BaseMessage;
+import org.openhab.binding.mycroft.internal.api.dto.MessageVolumeGet;
+import org.openhab.binding.mycroft.internal.channels.AudioPlayerChannel;
+import org.openhab.binding.mycroft.internal.channels.ChannelCommandHandler;
+import org.openhab.binding.mycroft.internal.channels.DuckChannel;
+import org.openhab.binding.mycroft.internal.channels.FullMessageChannel;
+import org.openhab.binding.mycroft.internal.channels.ListenChannel;
+import org.openhab.binding.mycroft.internal.channels.MuteChannel;
+import org.openhab.binding.mycroft.internal.channels.MycroftChannel;
+import org.openhab.binding.mycroft.internal.channels.SpeakChannel;
+import org.openhab.binding.mycroft.internal.channels.UtteranceChannel;
+import org.openhab.binding.mycroft.internal.channels.VolumeChannel;
+import org.openhab.core.io.net.http.WebSocketFactory;
+import org.openhab.core.thing.Channel;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.binding.BaseThingHandler;
+import org.openhab.core.types.Command;
+import org.openhab.core.types.State;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link MycroftHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Gwendal ROULLEAU - Initial contribution
+ */
+@NonNullByDefault
+public class MycroftHandler extends BaseThingHandler implements MycroftConnectionListener {
+
+    private final Logger logger = LoggerFactory.getLogger(MycroftHandler.class);
+
+    protected final MycroftConnection connection;
+    private @Nullable ScheduledFuture<?> scheduledFuture;
+    private MycroftConfiguration config = new MycroftConfiguration();
+    private boolean thingDisposing = false;
+    protected Map<ChannelUID, MycroftChannel<?>> mycroftChannels = new HashMap<>();
+
+    /** The reconnect frequency in case of error */
+    private static final int POLL_FREQUENCY_SEC = 30;
+    private int sometimesSendVolumeRequest = 0;
+
+    public MycroftHandler(Thing thing, WebSocketFactory webSocketFactory) {
+        super(thing);
+        String websocketID = thing.getUID().getAsString().replace(':', '-');
+        if (websocketID.length() < 4) {
+            websocketID = "mycroft-" + websocketID;
+        }
+        if (websocketID.length() > 20) {
+            websocketID = websocketID.substring(websocketID.length() - 20);
+        }
+        this.connection = new MycroftConnection(this, webSocketFactory.createWebSocketClient(websocketID));
```

Comment:
I would not create clients and connection objects here in the constructor. Better defer it to the `initialize()` method as it might be that the thing is disabled and in such a case there is no need to instantiate such objects at all.

Reply:
I didn't think about that. Thanks. Done